libnet/src/pf_key.rs

```diff
@@ -0,0 +1,830 @@
+/// creation.
+pub fn tcp_md5_key_update(
+    src: SockAddr,
+    dst: SockAddr,
+    authstring: &str,
+    valid_time: Duration,
+) -> Result<(), Error> {
+    tcp_md5_key_set(src, dst, authstring, valid_time, true)
+}
+
+/// Set a TCP-MD5 security association for the provided source and destination
+/// address with `authstring` as the key that is valid for `valid_time` after
+/// creation. If update is true, this is treated as an update to an existing
+/// association, otherwise a new association is created.
+pub fn tcp_md5_key_set(
+    src: SockAddr,
+    dst: SockAddr,
+    authstring: &str,
+    valid_time: Duration,
+    update: bool,
+) -> Result<(), Error> {
+    let msg =
+        TcpMd5SetKeyRequest::new(src, dst, authstring, valid_time, update);
+    let mut sock = Socket::new(
+        Domain::from(PF_KEY),
+        Type::RAW,
+        Some(Protocol::from(i32::from(PF_KEY_V2))),
+    )?;
+    let data = unsafe {
+        std::slice::from_raw_parts(
+            (&msg as *const TcpMd5SetKeyRequest) as *const u8,
+            size_of::<TcpMd5SetKeyRequest>(),
+        )
+    };
+    let n = sock.write(data)?;
+    if n != data.len() {
+        return Err(std::io::Error::new(
+            std::io::ErrorKind::UnexpectedEof,
+            format!("short write {} != {}", n, data.len()),
+        )
+        .into());
+    }
+
+    let mut buf = [0u8; 1024];
+    let _n = sock.read(&mut buf)?;
+    let response = unsafe { &*(buf.as_ptr() as *const Header) };
+
+    if response.errno != 0 {
+        return Err(Error::PfKey {
+            errno: response.errno,
+            typ: response.typ,
+            sa_typ: response.sa_typ,
+            diagnostic: response.reserved,
+        });
+    }
+
+    Ok(())
+}
+
+/// Get info on a TCP-MD5 security association for the provided source and
+/// destination address with
+pub fn tcp_md5_key_get(
+    src: SockAddr,
+    dst: SockAddr,
+) -> Result<GetAssociationResponse, Error> {
+    let msg = TcpMd5GetKeyRequest::new(src, dst);
+    let mut sock = Socket::new(
+        Domain::from(PF_KEY),
+        Type::RAW,
+        Some(Protocol::from(i32::from(PF_KEY_V2))),
+    )?;
+    let data = unsafe {
+        std::slice::from_raw_parts(
+            (&msg as *const TcpMd5GetKeyRequest) as *const u8,
+            size_of::<TcpMd5GetKeyRequest>(),
+        )
+    };
+    let n = sock.write(data)?;
+    if n != data.len() {
+        return Err(std::io::Error::new(
+            std::io::ErrorKind::UnexpectedEof,
+            format!("short write {} != {}", n, data.len()),
+        )
+        .into());
+    }
+
+    let mut buf = [0u8; 1024];
+    let _n = sock.read(&mut buf)?;
+    let response = unsafe { &*(buf.as_ptr() as *const Header) };
+
+    if response.errno != 0 {
+        return Err(Error::PfKey {
+            errno: response.errno,
+            typ: response.typ,
+            sa_typ: response.sa_typ,
+            diagnostic: response.reserved,
+        });
+    }
+
```

Comment:
Is it worth checking that the response has the expected sequence number?
You aren't setting the socket as promiscuous so you won't see SADB_EXPIRE messages that could be interspersed with the response so it is probably unnecessary to do that check.

Reply:
Updated the code check response sequence numbers.